src/cmdline/mod.rs

```diff
@@ -116,134 +97,98 @@ impl Cmdline {
     /// let cl = Cmdline::new(100);
     /// ```
     /// [`Cmdline`]: struct.Cmdline.html
-    pub fn new(capacity: usize) -> Cmdline {
-        assert_ne!(capacity, 0);
+    pub fn new(cmdline_capacity: usize) -> Cmdline {
+        assert!(cmdline_capacity > CMDLINE_NO_OF_AUXILIARY_CHARS);
+
         Cmdline {
-            line: String::with_capacity(capacity),
-            capacity,
+            boot_args: String::with_capacity(cmdline_capacity),
+            init_args: String::with_capacity(cmdline_capacity),
```

Comment:
This allows both components of the command line to have the maximum capacity, which would violate the capacity constraint on the entire `Cmdline` object. I don't think we can preallocate space for `String`s here anymore.

Reply:
If we remove the cmdline_capacity, the code is going to be greatly simplified. We could have a `MAX_SIZE` to make sure that the cmdline is still within reason to protect against potential malicious actors.

Reply:
I though about preallocating cmdline_capacity for both boot_args and init_args to simplify the implementation. Implementing like this guarantees that we have enough space even though no init arg will be provided (and thus, the entire space will be used for boot_args). We will NOT exceed cmdline_capacity when joining boot_args and init_args because before each insert in the Cmdline, we check the condition for capacity of the joined cmdline (the function has_capacity checks for the sum of lengths of boot_args and init_args). Yes, we can definitely implement without preallocations and keep the same checks - the only disadvantage that I can see for this approach would be the overhead introduced by repetitive inserts into string, but probably this would be insignificant as long as there are few inserts for the Cmdline. I wouldn't remove the cmdline_capacity from the constructor parameters as long as this is a constant specific to the architecture and I consider it part of the abstractions the rust-vmm/linux-loader imposes.